nemo_text_processing/text_normalization/normalize.py

```diff
@@ -12,198 +12,211 @@
-    if verbose:
-        print(text)
-        print(output)
-        print([str(tag) for tag in tags])
-    return output
 
-
-def normalize_identity(texts: List[str], verbose: bool = False) -> List[str]:
-    """
-    Identity normalizer. Returns input unchanged
-    Args:
-        texts: input string
-    Returns input string
+class Normalizer:
     """
-    return texts
-
+    Normalizer class that converts text from written to spoken form. 
+    Useful for TTS preprocessing. 
 
-def normalize_nemo(texts: List[str], verbose: bool = False) -> List[str]:
-    """
-    Text normalization with NeMo algorithm.
     Args:
-        texts: List of unnormlized strings
-        verbose: if specified prints debugging info
-    Returns list of normalized strings
-    """
-    res = []
-    for input in tqdm(texts):
-        text = normalize(input, verbose=verbose)
-        res.append(text)
-    return res
-
-
-normalizers = {"identity": normalize_identity, "nemo": normalize_nemo}
+        input_case: expected input capitalization
+    """
+
+    def __init__(self, input_case: str):
+        assert input_case in ["lower_cased", "cased"]
+
+        self.tagger = ClassifyFst(input_case=input_case)
+        self.verbalizer = VerbalizeFinalFst()
+        self.parser = TokenParser()
+
+    def normalize_list(self, texts: List[str], verbose=False) -> List[str]:
+        """
+        NeMo text normalizer 
+
+        Args:
+            texts: list of input strings
+
+        Returns converted list input strings
+        """
+        res = []
+        for input in tqdm(texts):
+            try:
+                text = self.normalize(input, verbose=verbose)
+            except:
+                print(input)
+                raise Exception
+            res.append(text)
+        return res
+
+    def normalize(self, text: str, verbose: bool) -> str:
+        """
+        Main function. Normalizes tokens from written to spoken form
+            e.g. 12 kg -> twelve kilograms
+
+        Args:
+            text: string that may include semiotic classes.
+
+        Returns: spoken form
+        """
+        text = text.strip()
+        if not text:
+            if verbose:
+                print(text)
+            return text
+        text = pynini.escape(text)
+        tagged_lattice = self.find_tags(text)
+        tagged_text = self.select_tag(tagged_lattice)
+        if verbose:
+            print(tagged_text)
+        self.parser(tagged_text)
+        tokens = self.parser.parse()
+        tags_reordered = self.generate_permutations(tokens)
+        for tagged_text in tags_reordered:
+            tagged_text = pynini.escape(tagged_text)
+            verbalizer_lattice = self.find_verbalizer(tagged_text)
+            if verbalizer_lattice.num_states() == 0:
+                continue
+            output = self.select_verbalizer(verbalizer_lattice)
+            return output
+        raise ValueError()
+
+    def _permute(self, d: OrderedDict) -> List[str]:
+        """
```

Comment:
can this func and most of others in this file  be reused in ITN/TN?

Reply:
I tried and its not trivial, imports mess up references. Can i try this in another PR?